### ISSUE TITLE
Setting question value to empty array is broken

### ISSUE THREAD

**Issue by SamMousa** [surveyjs/survey-library#6721]
### Are you requesting a feature, reporting a bug or asking a question?
Bug

### What is the current behavior?
A recent code change #6309 
https://github.com/surveyjs/survey-library/commit/fa65e2ee406d6d1af68fe47d363c78c7de766541#r124549343

added some "validation" when setting a question value. Now initializing a question without value (`undefined`) to an empty array `[]` no longer works.

### What is the expected behavior?
An empty array is not the same as `undefined` and calling `question.value = []` should work. 
I'd go even further and state that if it doesn't work an exception should be thrown.

While javascript supports getters and setters it is very unintuitive code if this does not hold:
```js
question.value = [];
if (question.value === []) {
   console.log('good, the setter works...');
}
```
If you really want to silently ignore this in a setter, an exception should be thrown.
By silencing errors instead of developers catching it during development, end users catch it when the code is already in production.

### How would you reproduce the current behavior (if this is a bug)?

Try setting the value of any question to `[]` and observe it doesn't work.

### Specify your

* surveyjs platform (angular or react or jquery or knockout or vue): knockout
* surveyjs version: 1.9.101

**Comment by andrewtelnov** [comment]
@SamMousa What question type do you want to set an empty array?
I have added several checks, because some developers can change the question type and after that they are trying to apply the previous entered data to a new question type. If a question expect an array and it is a string, they get an error. The opposite, when a question expect to have a string and get an array, the error is raising.

Thank you,
Andrew

**Comment by SamMousa** [comment]
It's a custom question type. But this is not about the wrong type of data, this is about an equality check that is incorrect.
Regardless; this hiding of errors is a bigger problem. You do not help developers by silently ignoring errors.

I do not want to do a check after every call to a SurveyJS library API...
```
survey.doSomething(); // Maybe it failed
if (survey.didYouReallyDoIt === false) {
    // do my error handling here
}
```

A developer needs to be able to reason about the state of a program. 
If I ask a 1000 developer to reason about the state of this program:
```js
question.value = 5;
// What is the value of question.value? ==> 5
question.value = [];
// What is the value of question.value? ==> []
```
There will be no developers that can reason their way towards `question.value` being `undefined` becaus it violates the basic assignment operator.

We might be having a philosophical difference of opinion here, but I'm confused as to how that is possible. Imagine every library you use implementing a different interpretation of basic operators, you'd never be able to get anything done.

**Comment by andrewtelnov** [comment]
@SamMousa value is a property. Question object assumbes that value, as all other properties, has a correct value. Otherwise the object should check is it correct or not before any operation with the property.
We have the following function that does the check:
```js
  protected isNewValueCorrect(val: any): boolean {
    return true;
  }
```
If the value is not correct we write a warning into console for a developer:
```js
    if(!this.isValueEmpty(newValue, !this.allowSpaceAsAnswer) && !this.isNewValueCorrect(newValue)) {
      ConsoleWarnings.inCorrectQuestionValue(this.name, newValue);
      return;
    }
```
I don't intend to break your code. I know how you feel when somebody breaks your code.
Let's create the fix where it works for you and we still have required checks. We have added this validation because developers complain that something doesn't work. However, the reason is incorrect question.value property. I personally, was tired of  debuging our customers projects and finally find this issue as a reason of problems.

You can override the `isNewValueCorrect` in your question type and return `true`. I can add an option into survey global settings variable to avoid value property checking. Please let me know, if a global settings is the best solution for you.

Thank you,
Andrew

**Comment by andrewtelnov** [comment]
@SamMousa Are you talking about this line:
`  if(!this.isValueEmpty(newValue, !this.allowSpaceAsAnswer) && !this.isNewValueCorrect(newValue)) {`
If your `isNewValueCorrect` always returns `true` then your code should work without any issue. Did I miss something?

Thank you,
Andrew

**Comment by andrewtelnov** [comment]
@SamMousa  Do you mean `isNewValueEqualsToValue` function? I believe that if checkbox has an empty array then the question is empty and the same for matrix dynamic and for panel dynamic.

Thank you,
Andrew

**Comment by SamMousa** [comment]
> I believe that if checkbox has an empty array then the question is empty and the same for matrix dynamic and for panel dynamic.

But what about a custom question type? Added via:
```js
    Survey.CustomWidgetCollection.Instance.addCustomWidget(widget, "customtype");
```

**Comment by andrewtelnov** [comment]
@SamMousa  I don't know your case. Probably empty array is a valid answer for your custom type. I have no idea. Please describe your case in more details.

Thank you,
Andrew

**Comment by SamMousa** [comment]
Yes it is a valid answer; otherwise I wouldn't be setting it... But the core code doesn't allow it and thinks it's equal to `undefined` and therefore doesn't set it...

**Comment by andrewtelnov** [comment]
@SamMousa By the way, what is your code? We have the following code in our library:
`const val = this.value || [];`. You can search for `|| []`.

Thank you,
Andrew

**Comment by andrewtelnov** [comment]
@SamMousa I see the issue. question.isEmpty() will return true for `undefined` and for empty array. 
Your code could be fixed as:
```js
const val = question.value || [];
//change val as you need
question.value = val;
```
We use this approach in our own code.

Thank you,
Andrew
